### PR TITLE
implements tree visitors

### PR DESCRIPTION
**Pull request by rbuckheit**
provides a way to operate on entire trees of parent/child objects

- includes example code of using a `DeletionVisitor` to delete a whole tree

### CODE REVIEW

File: js/src/tree/TreeLoader.coffee [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      @objectApi = new ObjectApi()
+
+    load: (id) ->
+      log.info('load', id)
+
+      return Promise.resolve()
+      .then =>
+        @objectApi.getObjectMetadata(id)
+      .then (metadata) =>
+        {childObjectCount} = metadata
```
Comment (makosblade) [inline js/src/tree/TreeLoader.coffee]: What does this {thing} syntax mean? I don't think I've ever seen this quite this way

File: demo/demo.js [hunk 2]
```diff
@@ -75,6 +80,13 @@ require([
       sharingClient.shareObject(objectId, shareWithUsers)
     });
 
+    // delete an object and all of its children resursively
```
Comment (sahiga) [inline demo/demo.js]: "recursively"

File: js/src/tree/TreeNode.coffee [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  #
+  class TreeNode
+
+    constructor: (@id, @children = []) ->
+      log.info('construct', {@id, @children})
+
+      validateId(@id)
+      validateChildren(@children)
+
+    # visits children depth-first and then the root note.
```
Comment (sahiga) [inline js/src/tree/TreeNode.coffee]: "node"

**Comment by sahiga** [comment]
Looks great overall :)

File: js/src/tree/TreeLoader.coffee (hunk 1 above)
Comment (rbuckheit) [inline js/src/tree/TreeLoader.coffee]: it's destructuring. for example if you have an object like:
`a = {id: '1', body: 'foo'}`

you can say
`{id, body} = a`

this compiles in to
```
var id = a.id
var body = a.body
```

File: demo/demo.js (hunk 2 above)
Comment (rbuckheit) [inline demo/demo.js]: lol, good call

File: js/src/tree/TreeNode.coffee (hunk 3 above)
Comment (rbuckheit) [inline js/src/tree/TreeNode.coffee]: thanks. we'll have to install a spell checking plugin for me :fearful:

**Pull request merged**